File: services/operations/rag-service/tracing.py
import time
import uuid
import json
import os

# Root trace file path matching Node.js
ARTIFACTS_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", "..", "..", "artifacts"))
TRACE_FILE = os.path.join(ARTIFACTS_DIR, "traces.json")

os.makedirs(ARTIFACTS_DIR, exist_ok=True)

class PythonSpan:
    def __init__(self, name, parent_context=None):
        self.name = name
        self.trace_id = parent_context.get("traceId") if parent_context else uuid.uuid4().hex
        self.span_id = uuid.uuid4().hex[:16]
        self.parent_span_id = parent_context.get("spanId") if parent_context else None
        self.start_time = time.perf_counter()
        self.attributes = {}
        self.events = []
        self.status = {"code": "UNSET"}

    def set_attribute(self, key, value):
        self.attributes[key] = value
        return self

    def add_event(self, name, attributes=None):
        self.events.append({
            "name": name,
            "time": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            "attributes": attributes or {}
        })
        return self

    def set_status(self, code, message=""):
        self.status = {"code": code, "message": message}
        return self

    def end(self):
        end_time = time.perf_counter()
        # Duration in milliseconds
        duration_ms = (end_time - self.start_time) * 1000.0

        span_data = {
            "name": self.name,
            "traceId": self.trace_id,
            "spanId": self.span_id,
            "parentSpanId": self.parent_span_id,
            "durationMs": round(duration_ms, 2),
            "startTimeStr": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
            "attributes": self.attributes,
            "events": self.events,
            "status": self.status
        }

        # Write thread-safe span output
        try:
            traces = []
            if os.path.exists(TRACE_FILE):
                try:
                    with open(TRACE_FILE, "r") as f:
                        traces = json.load(f)
                except Exception:
                    traces = []
            
            traces.append(span_data)
            with open(TRACE_FILE, "w") as f:
                json.dump(traces, f, indent=2)
        except Exception as e:
            print(f"[TRACING] Write error: {e}")

    @property
    def traceparent(self):
        return f"00-{self.trace_id}-{self.span_id}-01"


class PythonTracer:
    def __init__(self, service_name):
        self.service_name = service_name

    def start_span(self, name, traceparent=None):
        parent_context = None
        if traceparent and isinstance(traceparent, str):
            parts = traceparent.split("-")
            if len(parts) == 4:
                parent_context = {
                    "traceId": parts[1],
                    "spanId": parts[2]
                }
        
        span = PythonSpan(name, parent_context)
        span.set_attribute("service.name", self.service_name)
        return span

File: packages/shared-utils/tracing.js
/**
 * Light-Weight OpenTelemetry-Compatible SDK Layer
 *
 * Implements OpenTelemetry-compatible tracing API and trace propagation.
 * Automatically saves traces as a Directed Acyclic Graph (DAG) of spans,
 * allowing microsecond-level latency profiling and P99 bottleneck identification.
 */

const fs = require('fs');
const path = require('path');
const { v4: uuidv4 } = require('uuid');

const TRACE_FILE = path.join(__dirname, '..', '..', 'artifacts', 'traces.json');

// Ensure artifacts directory exists
try {
  fs.mkdirSync(path.dirname(TRACE_FILE), { recursive: true });
} catch (err) {}

class Span {
  constructor(name, parentContext = null) {
    this.name = name;
    this.traceId = parentContext ? parentContext.traceId : uuidv4().replace(/-/g, '');
    this.spanId = uuidv4().replace(/-/g, '').slice(0, 16);
    this.parentSpanId = parentContext ? parentContext.spanId : null;
    this.startTime = process.hrtime.bigint(); // microsecond resolution
    this.endTime = null;
    this.durationMs = null;
    this.attributes = {};
    this.events = [];
    this.status = { code: 'UNSET' };
  }

  setAttribute(key, value) {
    this.attributes[key] = value;
    return this;
  }

  addEvent(name, attributes = {}) {
    this.events.push({ name, time: new Date().toISOString(), attributes });
    return this;
  }

  setStatus(code, message = '') {
    this.status = { code, message };
    return this;
  }

  end() {
    this.endTime = process.hrtime.bigint();
    // Convert nanoseconds to milliseconds with float precision
    this.durationMs = Number(this.endTime - this.startTime) / 1e6;
    
    // Save span to shared file
    saveSpan(this);
  }

  get traceparent() {
    return `00-${this.traceId}-${this.spanId}-01`;
  }
}

function saveSpan(span) {
  const spanData = {
    name: span.name,
    traceId: span.traceId,
    spanId: span.spanId,
    parentSpanId: span.parentSpanId,
    durationMs: span.durationMs,
    startTimeStr: new Date().toISOString(),
    attributes: span.attributes,
    events: span.events,
    status: span.status,
  };

  try {
    let traces = [];
    if (fs.existsSync(TRACE_FILE)) {
      const content = fs.readFileSync(TRACE_FILE, 'utf8');
      traces = JSON.parse(content || '[]');
    }
    traces.push(spanData);
    fs.writeFileSync(TRACE_FILE, JSON.stringify(traces, null, 2));
  } catch (err) {
    // Fail-safe to prevent app crashes during disk contention
    console.error('[TRACING] Error writing span to trace file:', err.message);
  }
}

class Tracer {
  constructor(serviceName) {
    this.serviceName = serviceName;
  }

  /**
   * Starts a new span.
   * Can accept a W3C traceparent string to trace cross-service operations.
   */
  startSpan(name, options = {}) {
    let parentContext = null;
    const parent = options.parent || options.traceparent;

    if (parent && typeof parent === 'string') {
      const parts = parent.split('-');
      if (parts.length === 4) {
        parentContext = {
          traceId: parts[1],
          spanId: parts[2],
        };
      }
    } else if (parent && parent.traceId) {
      parentContext = parent;
    }

    const span = new Span(name, parentContext);
    span.setAttribute('service.name', this.serviceName);
    return span;
  }

  /**
   * Execute an async block inside a traced span
   */
  async trace(name, parent, fn) {
    const span = this.startSpan(name, { parent });
    try {
      const result = await fn(span);
      span.setStatus('OK');
      return result;
    } catch (err) {
      span.setStatus('ERROR', err.message);
      span.recordException ? span.recordException(err) : span.setAttribute('error', true);
      throw err;
    } finally {
      span.end();
    }
  }
}

/**
 * Audit traces to find P99 bottlenecks and build a DAG Call Graph
 */
function auditLatency() {
  if (!fs.existsSync(TRACE_FILE)) {
    return { error: 'No trace records found.' };
  }

  try {
    const spans = JSON.parse(fs.readFileSync(TRACE_FILE, 'utf8') || '[]');
    if (spans.length === 0) return { error: 'Trace file is empty.' };

    const latencyMap = {};
    const durations = [];

    spans.forEach(span => {
      durations.push(span.durationMs);
      if (!latencyMap[span.name]) {
        latencyMap[span.name] = [];
      }
      latencyMap[span.name].push(span.durationMs);
    });

    durations.sort((a, b) => a - b);
    const p99Idx = Math.floor(durations.length * 0.99);
    const p99Total = durations[p99Idx];

    const report = {
      totalSpansTracked: spans.length,
      globalP99Ms: p99Total.toFixed(2),
      spans: {}
    };

    Object.keys(latencyMap).forEach(name => {
      const list = latencyMap[name].sort((a, b) => a - b);
      const p99 = list[Math.floor(list.length * 0.99)];
      const avg = list.reduce((s, v) => s + v, 0) / list.length;
      report.spans[name] = {
        calls: list.length,
        avgLatencyMs: avg.toFixed(2),
        p99LatencyMs: p99.toFixed(2),
        bottleneckScore: p99 > 1000 ? 'HIGH 🚨' : (p99 > 200 ? 'MEDIUM ⚠️' : 'LOW ✅'),
      };
    });

    return report;
  } catch (err) {
    return { error: err.message };
  }
}

module.exports = {
  Tracer,
  auditLatency,
  TRACE_FILE
};
